src/Roslyn.Diagnostics.Analyzers/Core/SymbolIsBannedAnalyzer.cs

```diff
@@ -0,0 +1,231 @@
+
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationContext)
+        {
+            var additionalFiles = compilationContext.Options.AdditionalFiles;
+
+            if (!TryGetApiData(additionalFiles, compilationContext.CancellationToken, out ApiData apiData))
+            {
+                return;
+            }
+
+            if (!ValidateApiFiles(apiData, out List<Diagnostic> errors))
+            {
+                compilationContext.RegisterCompilationEndAction(context =>
+                {
+                    foreach (Diagnostic cur in errors)
+                    {
+                        context.ReportDiagnostic(cur);
+                    }
+                });
+
+                return;
+            }
+
+            var bannedApis = ImmutableHashSet.CreateBuilder<ISymbol>();
+            foreach (var apiString in apiData.ApiList)
+            {
+                var symbol = compilationContext.Compilation.GetTypeByMetadataName(apiString.Text);
+                if (!(symbol is null))
+                {
+                    bannedApis.Add(symbol);
+                }
+            }
+
+            if (bannedApis.Count > 0)
+            {
+                var symbolDisplayFormat = compilationContext.Compilation.Language == LanguageNames.CSharp
+                    ? SymbolDisplayFormat.CSharpShortErrorMessageFormat
+                    : SymbolDisplayFormat.VisualBasicShortErrorMessageFormat;
+                compilationContext.RegisterOperationAction(
+                    oac => AnalyzeOperation(oac, bannedApis.ToImmutable(), symbolDisplayFormat),
+                    OperationKind.ObjectCreation,
+                    OperationKind.Invocation,
+                    OperationKind.EventReference,
+                    OperationKind.FieldReference,
+                    OperationKind.MethodReference,
+                    OperationKind.PropertyReference);
+            }
+        }
+
+        private static void AnalyzeOperation(OperationAnalysisContext oac, ImmutableHashSet<ISymbol> bannedSymbols, SymbolDisplayFormat symbolDisplayFormat)
+        {
+            ITypeSymbol type = null;
+            switch (oac.Operation)
+            {
+                case IObjectCreationOperation objectCreation:
+                    type = objectCreation.Type.OriginalDefinition;
+                    break;
+
+                case IInvocationOperation invocation:
+                    type = invocation.TargetMethod.ContainingType.OriginalDefinition;
+                    break;
+
+                case IMemberReferenceOperation memberReference:
+                    type = memberReference.Member.ContainingType.OriginalDefinition;
+                    break;
+            }
+
+            if (!(type is null) && bannedSymbols.Contains(type))
```

Comment:
Why just types? It seems reasonable to expected even members, especially methods, to be banned based on performance characteristics.

Reply:
Because I was too lazy to implement resolution for members.  I thought @mattwar wrote some code to resolve from an xml doc comment id, but I couldn't find it on Saturday.  @sharwell also said that VSThreading analyzers faces a similar problem.

Reply:
I presume you meant this API: http://source.roslyn.io/#Microsoft.CodeAnalysis/DocumentationCommentId.cs,45. I was actually suggesting keeping the entries [consistent](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Roslyn.Diagnostics.Analyzers/UnitTests/DeclarePublicAPIAnalyzerTests.cs#L293) with existing public API files, but that sounds like an interesting approach as well. We use the DocumentationCommendId for suppress message attribute target string.

Reply:
That's the one.  Doc comment ids are already somewhat familiar to users, which is why I wanted to use that.  I'd hate to spread use of yet another string form of symbols on the world.

@tmeschter do you remember why the above wasn't sufficient for the public API analyzer (I think return type was one reason, not sure if there were others).

Reply:
The return type is certainly an issue, but I think the real reason is that the `DocumentationCommentId` either wasn't available to me when I wrote the Public API analyzer or I just didn't think to use it.

Reply:
The code's now using doc comment IDs. They're not the friendliest format to write for the uninitiated. We might be able to make this easier with another diagnostic/codefix.

Reply:
The format here now allows specifying a message to include in the diagnostic after a semicolon (@tmeschter expressed interest in `#` but that's reserved in doc comment IDs already).